File: versions/dsa/_codecheck/ch09_图论算法_cpp/lab.cpp
// lab9.cpp —— 迷宫最短路径(BFS + Dijkstra)
// 编译: g++ -std=c++17 -O2 -o lab9 lab9.cpp
// 运行: ./lab9          (内置迷宫)
//       ./lab9 sealed   (内置无解样例:S 被墙封死)
//       ./lab9 maze.txt (从文件读迷宫)
#include <algorithm>
#include <fstream>
#include <functional>
#include <iostream>
#include <queue>
#include <string>
#include <utility>
#include <vector>
using namespace std;

// 任务 1 辅助:字符 -> 通过代价(墙返回 -1)
int costOf(char c) {
    switch (c) {
        case 'S': case 'E': case '.': return 1;   // 平地
        case '~': return 5;                       // 沼泽:代价高
        default:  return -1;                      // '#' 墙
    }
}

vector<string> readMaze(const string& arg) {
    if (arg == "sealed") {                        // 内置样例 2:S 被墙封死
        return {"#####", "##S##", "#.###", "#.E.#", "#####"};
    }
    if (!arg.empty()) {                           // 从文件读
        ifstream in(arg);
        vector<string> m;
        string line;
        while (getline(in, line))
            if (!line.empty()) m.push_back(line);
        return m;
    }
    return {                                      // 内置样例 1:
        "##########",                             // 上走廊短(12 步)但有两格
        "#S...~~..#",                             // 沼泽(代价 21);下走廊长
        "#.#####.##",                             // (18 步)但全平地(代价 18)
        "#.#####.##",                             // ——BFS 选短,Dijkstra 选廉
        "#.##....##",
        "#.##.##.##",
        "#.##.##.##",
        "#....##E##",
        "##########",
        "##########",
    };
}

// 任务 1:网格 -> 图,顶点编号 id = 行 * 宽 + 列,四方向邻居即边(隐式图)
int findPos(const vector<string>& maze, char ch) {
    int w = int(maze[0].size());
    for (int r = 0; r < int(maze.size()); ++r)
        for (int c = 0; c < w; ++c)
            if (maze[r][c] == ch) return r * w + c;
    return -1;
}

// 任务 2:无权最短路(BFS)。返回步数;无解返回 -1
int bfsShortest(const vector<string>& maze, int sid, int tid, vector<int>& path) {
    int h = int(maze.size()), w = int(maze[0].size());
    vector<int> dist(h * w, -1), prev(h * w, -1);
    queue<int> q;                                 // 队列(第3章)
    dist[sid] = 0;
    q.push(sid);
    while (!q.empty()) {
        int u = q.front(); q.pop();
        if (u == tid) break;
        int r = u / w, c = u % w;
        int dr[4] = {-1, 1, 0, 0}, dc[4] = {0, 0, -1, 1};
        for (int k = 0; k < 4; ++k) {
            int nr = r + dr[k], nc = c + dc[k];
            if (nr < 0 || nr >= h || nc < 0 || nc >= w) continue;
            if (maze[nr][nc] == '#') continue;    // 墙不可走
            int v = nr * w + nc;
            if (dist[v] != -1) continue;          // 第一次到达即最短
            dist[v] = dist[u] + 1;
            prev[v] = u;
            q.push(v);
        }
    }
    if (dist[tid] == -1) return -1;               // 任务 5:终点不可达
    for (int x = tid; x != -1; x = prev[x])       // 前驱数组回溯
        path.push_back(x);
    reverse(path.begin(), path.end());
    return int(path.size()) - 1;
}

// 任务 3:最低代价路径(Dijkstra 堆优化)。返回总代价;无解返回 -1
int dijkstraCost(const vector<string>& maze, int sid, int tid, vector<int>& path) {
    const int INF = 1e9;                          // 大常量,INF + w 不溢出
    int h = int(maze.size()), w = int(maze[0].size());
    vector<int> dist(h * w, INF), prev(h * w, -1);
    priority_queue<pair<int, int>, vector<pair<int, int>>,
                   greater<pair<int, int>>> pq;   // 最小堆(第6章)
    dist[sid] = 0;
    pq.push({0, sid});
    while (!pq.empty()) {
        auto [d, u] = pq.top(); pq.pop();
        if (d > dist[u]) continue;                // 懒删除:过期记录直接跳过
        if (u == tid) break;
        int r = u / w, c = u % w;
        int dr[4] = {-1, 1, 0, 0}, dc[4] = {0, 0, -1, 1};
        for (int k = 0; k < 4; ++k) {
            int nr = r + dr[k], nc = c + dc[k];
            if (nr < 0 || nr >= h || nc < 0 || nc >= w) continue;
            int wgt = costOf(maze[nr][nc]);
            if (wgt < 0) continue;                // 墙
            int v = nr * w + nc;
            if (dist[v] > d + wgt) {
                dist[v] = d + wgt;
                prev[v] = u;
                pq.push({dist[v], v});
            }
        }
    }
    if (dist[tid] == INF) return -1;
    for (int x = tid; x != -1; x = prev[x])
        path.push_back(x);
    reverse(path.begin(), path.end());
    return dist[tid];
}

void drawPath(const vector<string>& maze, const vector<int>& path, int w) {
    vector<string> grid = maze;
    for (int id : path)
        if (grid[id / w][id % w] != 'S' && grid[id / w][id % w] != 'E')
            grid[id / w][id % w] = '*';           // 用 * 标出路径
    for (const string& row : grid) {
        for (char ch : row) cout << ' ' << ch;
        cout << '\n';
    }
}

void printPathCoords(const vector<int>& path, int w) {
    for (size_t i = 0; i < path.size(); ++i)
        cout << (i ? " -> " : "")
             << "(" << path[i] / w << "," << path[i] % w << ")";
    cout << '\n';
}

int main(int argc, char* argv[]) {
    string arg = argc > 1 ? argv[1] : "";
    vector<string> maze = readMaze(arg);
    int w = int(maze[0].size());
    int s = findPos(maze, 'S'), t = findPos(maze, 'E');
    if (s < 0 || t < 0) {
        cout << "迷宫缺少起点 S 或终点 E\n";
        return 1;
    }
    cout << "迷宫 " << maze.size() << " 行 x " << w << " 列,起点 ("
         << s / w << "," << s % w << "),终点 (" << t / w << "," << t % w << ")\n";
    cout << "图例: # 墙  . 平地(代价1)  ~ 沼泽(代价5)  S/E 起点/终点\n";

    // 任务 2:BFS 无权最短路
    vector<int> pathB;
    int steps = bfsShortest(maze, s, t, pathB);
    if (steps < 0) {
        cout << "BFS  : 无解——起点到终点不可达(被墙封死)\n";
    } else {
        cout << "BFS  : 最短步数 = " << steps << "\n  路径: ";
        printPathCoords(pathB, w);
        drawPath(maze, pathB, w);
    }

    // 任务 3:Dijkstra 最低代价路径
    vector<int> pathD;
    int cost = dijkstraCost(maze, s, t, pathD);
    if (cost < 0) {
        cout << "Dijkstra: 无解\n";
    } else {
        cout << "Dijkstra: 最低总代价 = " << cost
             << ",步数 = " << int(pathD.size()) - 1 << "\n  路径: ";
        printPathCoords(pathD, w);
        drawPath(maze, pathD, w);
    }

    // 任务 4:两种路径对比
    if (steps >= 0 && cost >= 0) {
        int bfsCost = 0;                          // BFS 路径的真实代价
        for (int id : pathB) bfsCost += costOf(maze[id / w][id % w]);
        cout << "对比: BFS 最短步数 " << steps << "(路径代价 " << bfsCost
             << ") vs Dijkstra 最低代价路径步数 " << int(pathD.size()) - 1
             << "(代价 " << cost << ")\n";
    }
    return 0;
}

File: versions/dsa/_codecheck/ch09_图论算法_py/lab.py
# lab9.py —— 迷宫最短路径(BFS + Dijkstra)
# 运行: python lab9.py            # 内置迷宫
#       python lab9.py sealed     # 内置"无解"迷宫(S 被墙封死)
#       python lab9.py maze.txt   # 从文件读迷宫
import heapq
import sys
from collections import deque

# 内置迷宫:'#' 墙,'.' 平地(代价 1),'~' 沼泽(代价 5),'S' 起点,'E' 终点
# 设计: 上走廊短(12 步)但有两格沼泽(代价 21);
#       下走廊长(18 步)但全平地(代价 18)——BFS 选短,Dijkstra 选廉。
MAZE = """\
##########
#S...~~..#
#.#####.##
#.#####.##
#.##....##
#.##.##.##
#.##.##.##
#....##E##
##########
##########\
"""

SEALED = """\
#####
##S##
#.###
#.E.#
#####\
"""

COST = {'S': 1, 'E': 1, '.': 1, '~': 5}


def read_maze(arg):
    if arg == "sealed":
        return SEALED.splitlines()
    if arg:
        with open(arg, encoding="utf-8") as f:
            return [line.rstrip("\n") for line in f if line.strip()]
    return MAZE.splitlines()


def find_pos(maze, ch):
    for r, row in enumerate(maze):
        for c, cell in enumerate(row):
            if cell == ch:
                return r, c
    return -1, -1


def neighbors(maze, r, c):
    """任务 1:网格 -> 图,四方向邻居即边(隐式图)。"""
    h, w = len(maze), len(maze[0])
    for dr, dc in ((-1, 0), (1, 0), (0, -1), (0, 1)):
        nr, nc = r + dr, c + dc
        if 0 <= nr < h and 0 <= nc < w and maze[nr][nc] != '#':
            yield nr, nc


def bfs_shortest(maze, s, t):
    """任务 2:无权最短路(BFS)。返回 (步数, 路径坐标列表);无解返回 None。"""
    q = deque([s])
    prev = {s: None}
    while q:
        u = q.popleft()
        if u == t:
            break
        for v in neighbors(maze, *u):
            if v not in prev:          # 第一次到达即最短
                prev[v] = u
                q.append(v)
    if t not in prev:
        return None                    # 任务 5:终点不可达
    path = []
    x = t
    while x is not None:               # 前驱数组回溯
        path.append(x)
        x = prev[x]
    path.reverse()
    return len(path) - 1, path


def dijkstra_cost(maze, s, t):
    """任务 3:最低代价路径(Dijkstra 堆优化)。返回 (总代价, 路径);无解返回 None。"""
    h, w = len(maze), len(maze[0])
    INF = float('inf')                 # Python 用无穷大,无溢出问题
    dist = [[INF] * w for _ in range(h)]
    prev = {}
    dist[s[0]][s[1]] = 0
    pq = [(0, s)]                      # (当前代价, 坐标),小顶堆
    while pq:
        d, u = heapq.heappop(pq)
        if d > dist[u[0]][u[1]]:
            continue                   # 懒删除:过期记录直接跳过
        if u == t:
            break
        for v in neighbors(maze, *u):
            nd = d + COST[maze[v[0]][v[1]]]
            if nd < dist[v[0]][v[1]]:
                dist[v[0]][v[1]] = nd
                prev[v] = u
                heapq.heappush(pq, (nd, v))
    if dist[t[0]][t[1]] == INF:
        return None
    path = []
    x = t
    while x is not None:
        path.append(x)
        x = prev.get(x)
    path.reverse()
    return dist[t[0]][t[1]], path


def draw_path(maze, path):
    """把路径画回迷宫:* 标出经过的格子。"""
    grid = [list(row) for row in maze]
    for r, c in path:
        if grid[r][c] not in 'SE':
            grid[r][c] = '*'
    return [' '.join(row) for row in grid]


def main():
    arg = sys.argv[1] if len(sys.argv) > 1 else ""
    maze = read_maze(arg)
    s, t = find_pos(maze, 'S'), find_pos(maze, 'E')
    if s == (-1, -1) or t == (-1, -1):
        print("迷宫缺少起点 S 或终点 E")
        return
    print(f"迷宫 {len(maze)} 行 x {len(maze[0])} 列,起点 {s},终点 {t}")
    print("图例: # 墙  . 平地(代价1)  ~ 沼泽(代价5)  S/E 起点/终点")

    res_bfs = bfs_shortest(maze, s, t)
    if res_bfs is None:
        print("BFS  : 无解——起点到终点不可达(被墙封死)")
    else:
        steps, path = res_bfs
        print(f"BFS  : 最短步数 = {steps}")
        print("  路径:", " -> ".join(f"({r},{c})" for r, c in path))
        for row in draw_path(maze, path):
            print("   ", row)

    res_dij = dijkstra_cost(maze, s, t)
    if res_dij is None:
        print("Dijkstra: 无解")
    else:
        cost, path = res_dij
        print(f"Dijkstra: 最低总代价 = {cost},步数 = {len(path)-1}")
        print("  路径:", " -> ".join(f"({r},{c})" for r, c in path))
        for row in draw_path(maze, path):
            print("   ", row)

    # 任务 4:两种路径对比
    if res_bfs and res_dij:
        steps, path = res_bfs
        cost, path2 = res_dij
        bfs_cost = sum(COST[maze[r][c]] for r, c in path)   # BFS 路径的真实代价
        print(f"对比: BFS 最短步数 {steps}(路径代价 {bfs_cost}) vs "
              f"Dijkstra 最低代价路径步数 {len(path2)-1}(代价 {cost})")


if __name__ == "__main__":
    main()
